### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] Maximum call stack size exceeded

### ISSUE THREAD

**Issue by TimKolberger** [typescript-eslint/typescript-eslint#9705]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=KYDwDg9gTgLgBDAnmYcCCA7AhjCBbRASQwGcYsMBjYAHgBU5QZgMATEuMqASwwHMAfHAC8cAPJ5uMGphz5E9AQBo4AcjBZYJVQIBQu0JFhxezKADMs1dNlwF6jEMzYcuvQXADeuuI%2BesAdSkACwAuOBoANT8Wdk4YHn4BAAoAOnSNLXDIgG0AXQBKESFZOyJScipaBgAfOEi9AF9dIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
export type AnatomyInstance<T extends string> = Omit<Anatomy<T>, 'parts'>

export interface Anatomy<T extends string> {
  extendWith: <V extends string>(...parts: V[]) => AnatomyInstance<T | V>
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error

### Actual Result

Maximum call stack size exceeded

**Comment by kirkwaiblinger** [comment]
What fun!

**Comment by kirkwaiblinger** [comment]
Breadcrumb for investigation...

From 2 seconds of playing around - I'm think this is triggered by something to do with the circular dependency between the `Anatomy` and `AnatomyInstance` types.

**Comment by bradzacher** [comment]
It probably committed the common problem of not short-circuiting for already seen types. We do this all the time heh.

**Comment by abrahamguo** [comment]
So far, what I've found out is that, of course, it is due to infinite recursion through already-seen types. However, `no-unnecessary-type-parameters` _does_ already have protection against infinite recursion. The problem, though, is that according to printing out the chain of recursion using `checker.typeToString()`, we enter the following infinite loop:

```
<V extends string>(...parts: V[]) => AnatomyInstance<V | T>
AnatomyInstance<V | T>
<V extends string>(...parts: V[]) => AnatomyInstance<V | T | V>
AnatomyInstance<V | T | V>
<V extends string>(...parts: V[]) => AnatomyInstance<V | T | V | V>
AnatomyInstance<V | T | V | V>
<V extends string>(...parts: V[]) => AnatomyInstance<V | T | V | V | V>
```

Because additional type parameters are added, of course, it does not trigger the infinite recursion check. I still have to determine why that is.

**Comment by abrahamguo** [comment]
Good news — looks like this was actually fixed in #9741!

**Comment by kirkwaiblinger** [comment]
> Good news — looks like this was actually fixed in #9741!

Woot! 🎉
